src/Compilers/Test/Core/Assert/AssertEx.cs

```diff
@@ -594,37 +616,117 @@ public static void AssertResultsEqual(string result1, string result2)
+        internal static ReadOnlyMemory<char> NormalizeWhitespaceROM(ReadOnlySpan<char> input, out char[] pooledAlloc)
         {
-            var output = new StringBuilder();
-            var inputLines = input.Split('\n', '\r');
-            foreach (var line in inputLines)
+            // Setup allocation (start with none)
+            pooledAlloc = null;
+            int allocSize = 0;
+
+            // Loop over each line section until we run out of lines
+            int prev = 0;
+            int next = input.IndexOfAny("\r\n".AsSpan());
+            if (next < 0) next = input.Length;
+            while (true)
             {
-                var trimmedLine = line.Trim();
+                // Get the line, trimmed
+                var trimmedLine = input.Slice(prev, next - prev).Trim();
                 if (trimmedLine.Length > 0)
                 {
+                    // We want to append 2 spaces first if we don't start with { or }
+                    int spacesToAppend = 0;
                     if (!(trimmedLine[0] == '{' || trimmedLine[0] == '}'))
                     {
-                        output.Append("  ");
+                        spacesToAppend = 2;
                     }
 
-                    output.AppendLine(trimmedLine);
+                    // Ensure buffer size is big enough
+                    int newChars = spacesToAppend + trimmedLine.Length + Environment.NewLine.Length;
+                    pooledAlloc ??= ArrayPool<char>.Shared.Rent(newChars);
+                    if (pooledAlloc.Length < allocSize + newChars)
+                    {
+                        char[] newArr = ArrayPool<char>.Shared.Rent(allocSize + newChars);
+                        pooledAlloc.AsSpan().Slice(0, allocSize).CopyTo(newArr.AsSpan().Slice(0, allocSize));
+                        ArrayPool<char>.Shared.Return(pooledAlloc);
+                        pooledAlloc = newArr;
+                    }
+
+                    // Copy new text into buffer
+                    pooledAlloc.AsSpan().Slice(allocSize, spacesToAppend).Fill(' ');
+                    allocSize += spacesToAppend;
+                    trimmedLine.CopyTo(pooledAlloc.AsSpan().Slice(allocSize, trimmedLine.Length));
+                    allocSize += trimmedLine.Length;
+                    Environment.NewLine.AsSpan().CopyTo(pooledAlloc.AsSpan().Slice(allocSize, Environment.NewLine.Length));
+                    allocSize += Environment.NewLine.Length;
                 }
+
+                // Exit if done
+                if (next == input.Length) break;
+
+                // Find next section
+                prev = next + 1;
+                next = input.Slice(prev).IndexOfAny("\r\n".AsSpan());
+                if (next >= 0) next += prev;
+                else next = input.Length;
             }
 
-            return output.ToString();
+            // Return the valid part of our allocation
+            if (pooledAlloc == null) return default;
```

Comment:
Nit: the style in roslyn requires that this be put on separate lines. There are a few instances of this in the PR that need attention.